src/main/java/com/npe/pet/uncaughtHeroes/service/HeroService.java

```diff
@@ -1,36 +1,45 @@
 package com.npe.pet.uncaughtHeroes.service;
 
 import com.npe.pet.uncaughtHeroes.entity.Hero;
+import com.npe.pet.uncaughtHeroes.exception.HeroNameDuplicateException;
+import com.npe.pet.uncaughtHeroes.exception.HeroNotFoundException;
 import com.npe.pet.uncaughtHeroes.repository.HeroRepository;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.DuplicateKeyException;
 import org.springframework.stereotype.Service;
 
 import java.util.List;
 import java.util.Optional;
 
 @Service
 public class HeroService {
+    private static final String HERO_NAME_DUPLICATE_MESSAGE = "'%s' hero already exists.";
+    private static final String HERO_NOT_FOUND_MESSAGE = "Hero was not found. (ID: %s)";
     private final HeroRepository heroRepository;
 
-    @Autowired
     public HeroService(HeroRepository heroRepository) {
         this.heroRepository = heroRepository;
     }
 
     public Hero save(Hero hero) {
-        return heroRepository.save(hero);
+        try {
+            return heroRepository.save(hero);
+        } catch (DuplicateKeyException exception) {
+            throw new HeroNameDuplicateException(String.format(HERO_NAME_DUPLICATE_MESSAGE, hero.getName()));
```

Comment:
I would extract the message generation into private methods:

like private String getDuplicateMessage(Hero hero) {
   ...
}